sdk/core/Azure.Core/src/Serialization/ModelWriter.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core.Serialization
+{
+    /// <summary>
+    /// Provides an efficient way to serialize <see cref="IModelJsonSerializable{T}"/> into a <see cref="BinaryData"/> using multiple pooled buffers.
+    /// </summary>
+    internal sealed partial class ModelWriter : IDisposable
```

Comment:
Curious (trying to understand) - it looks like this class implements a wrapper around SequenceBuilder, almost exclusively, and adds a layer synchronizing access to the SequenceBuilder.  Is that correct?  Does every model call through to it, via ModelSerializer.SerializerCore()?  Why is synchronized access needed?

Reply:
yes every model calls through this and yes your understanding is correct this is a thread sync wrapper around sequence builder.

if we don't have this protection sequence builder can expose use after free issues.  we plan to optimize this in the future but with this implementation we get the perf boots of sequence builder without exposing that possibility.